Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,212 @@
+
+The view-distance config value will determine the global distance (and set maximum) of normal view distance interactions with entities, mob-spawning and tracking. All chunks with player tickets within this range will also be served by the server, allowing shared view distance from other players without impacting performance.
+
+diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+index 63a688725..78a69973a 100644
+--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
++++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+@@ -20,6 +20,7 @@ import it.unimi.dsi.fastutil.objects.ObjectIterator;
+ import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+ import it.unimi.dsi.fastutil.objects.ObjectSet;
+ import it.unimi.dsi.fastutil.objects.ObjectSortedSet;
++
+ import java.util.Set;
+ import java.util.concurrent.CompletableFuture;
+ import java.util.concurrent.Executor;
+@@ -44,6 +45,8 @@ public abstract class ChunkMapDistance {
+     private final Executor m;
+     private long currentTick;
+ 
++    protected ObjectSet<EntityPlayer> getChunkPlayers(long i) {return (ObjectSet) ChunkMapDistance.this.c.get(i);} //Paper OBFHELPER
++
+     protected ChunkMapDistance(Executor executor, Executor executor1) {
+         executor1.getClass();
+         Mailbox<Runnable> mailbox = Mailbox.a("player ticket throttler", executor1::execute);
+@@ -352,7 +355,10 @@ public abstract class ChunkMapDistance {
+ 
+         private int e = 0;
+         private final Long2IntMap f = Long2IntMaps.synchronize(new Long2IntOpenHashMap());
+-        private final LongSet g = new LongOpenHashSet();
++        private final Long2ByteMap g = new Long2ByteOpenHashMap(); //Paper ticket level pp-view-dist
++
++        private ObjectIterator<Long2ByteMap.Entry> getUpdateIterator() {return this.g.long2ByteEntrySet().iterator();} //Paper OBFHELPER
++        private void insertUpdate(long i, byte d) {this.g.put(i, d);} //Paper OBFHELPER
+ 
+         protected c(int i) {
+             super(i);
+@@ -360,9 +366,9 @@ public abstract class ChunkMapDistance {
+         }
+ 
+         @Override
+-        protected void a(long i, int j, int k) {
+-            this.g.add(i);
+-        }
++        protected void a(long i, int j, int k, byte d) {
++            this.insertUpdate(i, d);
++        } //Paper ticket level pp-view-dist
+ 
+         public void a(int i) {
+             ObjectIterator objectiterator = this.a.long2ByteEntrySet().iterator();
+@@ -372,20 +378,20 @@ public abstract class ChunkMapDistance {
+                 byte b0 = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getByteValue();
+                 long j = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getLongKey();
+ 
+-                this.a(j, b0, this.c(b0), b0 <= i - 2);
++                this.a(j, b0, this.c(b0), b0 <= i - 2, i); //Paper ticket level pp-view-dist
```

Comment:
this call is still checking against the global view distance

Reply:
this is intentional, this method is only really called once on world initialization, and attempting to pass accurate per player information through here would be rather unnecessary and risky.